### ISSUE TITLE
Issues with JBoss/TomEE to set Container-managed Transactions with Kundera

### ISSUE THREAD

**Issue by emgsilva** [Impetus/kundera#344]
Hi,

After some insights from Vivek on the other day on the benefits of having “Container-managed EntityManagers” I spend some time trying to port our Kundera persistences to such a setup. However I am struggling with several issues… hopefully you can give me some insights. I do not have much experience with this (EJBs, Java EE containers, etc.), but I tried to learn some basics and set some experiments running… Basically I use the “@PersistenceContext” to inject the EntityManager corresponding to the Persistence Unit I want to handle the persistence (declared in Kundera persistence.xml)… this injection takes place on a “@Stateless” EJB… I thought this would be enough… but something else probably needs to be done.

I am experimenting two different Java EE Application Servers: JBoss AS 7.1.1 and TomEE 1.5.2, both give me errors (different types of errors - Jboss: https://www.dropbox.com/s/fuzjq9af6qenq7w/jboss-problem.txt  | TomEE:  https://www.dropbox.com/s/qs3ilax3i1tuse9/tomee-problem.txt).

My experiments consisted of trying to call one of my (RESTful web service) Server API and in the class handling the communications with the DB (Cassandra using Kundera) I acquire the EntityManager with the following code: 

```
...
@Stateless
public class CassandraDomainDataManager{

     @PersistenceContext(unitName="cassandra_ServiceTypes")
     private EntityManager emServiceTypes;
...
```

I then do other operations with the DB in another class/bean… but I have also problems (as you can seen in the TomEE logs).

In fact I only manage to get to tests with TomEE, with JBoss I cannot even fully start the server (even including the recommendations from Vivek - https://github.com/impetus-opensource/Kundera/issues/180#issuecomment-22089687)… 

This is an excerpt of my Kundera persistence.xml:

```
<persistence xmlns="http://java.sun.com/xml/ns/persistence"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="http://java.sun.com/xml/ns/persistence
     http://java.sun.com/xml/ns/persistence/persistence_2_0.xsd"
     version="2.0”>

    <persistence-unit name="cassandra_ServiceTypes" transaction-type="JTA">
          <provider>com.impetus.kundera.KunderaPersistence</provider>
          <jta-data-source>CassandraDS</jta-data-source>
          <class>com.sigmabees.data.cassandra.servtypes.ServiceTypes</class>
          <class>com.sigmabees.data.cassandra.servtypes.ClassificationProperties</class>
          <class>com.sigmabees.data.cassandra.servtypes.ClassificationPropertiesValues</class>
          <class>com.sigmabees.data.cassandra.servtypes.InstanceSpecificationProperties</class>
          <class>com.sigmabees.data.cassandra.servtypes.Domains</class>
          <class>com.sigmabees.data.cassandra.servtypes.OfferDocumentationProperties</class>
          <class>com.sigmabees.data.cassandra.servtypes.InputSemTypes</class>
          <exclude-unlisted-classes>true</exclude-unlisted-classes>
          <properties>
               <property name="kundera.nodes" value="localhost" />
               <property name="kundera.port" value="9160" />
               <property name="kundera.keyspace" value="servicetypes" />
               <property name="kundera.dialect" value="cassandra" />
               <property name="kundera.client" value="pelops" />
               <property name="kundera.client.lookup.class"
                    value="com.impetus.client.cassandra.thrift.ThriftClientFactory" />
               <property name="kundera.cache.provider.class"
                    value="com.impetus.kundera.cache.ehcache.EhCacheProvider" />
               <property name="kundera.cache.config.resource" value="/ehcache-test.xml" />
               <property name="kundera.pool.size.max.active" value=“50"></property>
               <property name="jboss.as.jpa.managed" value="false" />
          </properties>
     </persistence-unit>
...
```

I am doing something wrong… would appreciate some insights if you have some, mainly if this is something related with Kundera configurations, or it is something related with my bad handling of the EJBs/Containers/etc.

Cheers,

Eduardo

**Comment by emgsilva** [comment]
I forgot to mention... in TomEE the problem (NullExceptionPointer) is obtained when I try to access the EntityManager, it is "null", so somehow it is not being "injected":

```
...
@Stateless
public class CassandraDomainDataManager{

     @PersistenceContext(unitName="cassandra_ServiceTypes")
     private EntityManager emServiceTypes;

     public boolean checkDomainSupport(String domain){

             Domains domainData = emServiceTypes.find(Domains.class, domain); // NEP
...
```

**Comment by mevivs** [comment]
Could you please email me those error logs at vivek.mishra@impetus.co.in? As i am not able to access dropbox links from office.

-Vivek

**Comment by emgsilva** [comment]
This problem was solved (thanks to precious help from Vivek!).

I made several adjustments on my code and persistence.xml configuration:
- 1) created "stateless local beans", which work as "facades" to communications with Cassandra DBs via Kundera. In these beans I declare/inject the EntityManager (@PersistenceContext(unitName="cassandra_ServiceTypes", type=PersistenceContextType.TRANSACTION), see the following example... then in this bean I perform several operations with Kundera:

```
@Stateless
@Local
public class CassandraDomainDataManager{

    @PersistenceContext(unitName="cassandra_ServiceTypes", type=PersistenceContextType.TRANSACTION)
    private EntityManager emServiceTypes;

    // methods making use of EntityManager (emServiceTypes)
```
- 2) in my REST API, or whenever required, I inject a @EJB when instantiating CassandraDomainDataManager. In the class making that call I also use a "@Stateless" bean declaration. This instance can then be used to make all the calls to the CassandraDomainDataManager methods... See example in the following:

```
@Stateless
@Path("/servtypes")
@Produces({ "application/json"})
public class ServiceTypeAPI{

    @EJB
    private CassandraDomainDataManager cassandraDomainDataManager =
            new CassandraDomainDataManager();
```
- 3) I removed all "JTA" references from my persistency.xml persistence-unit declarations... Kundera and other NoSQL DBs do not use/require that.

This was it more or less... Soon more detailed documentation will be posted in Kundera wiki about Container-managed EntityManager (in Java EE Application Servers).

Cheers!
